### PR TITLE
VideoCommon: Add dynamic vertex loader for ubershaders to reduce pipeline count

### PR DESCRIPTION
**Pull request by tellowkrinkle**
Some fun things I've been working on, figured I'd post it here for other people to try out, even if it's not quite ready for merge

Adds a vertex loader to vertex ubershaders which removes the need to do anything with the input assembler, reducing the number of unique pipelines needed.  This makes the number of different pipelines that actually require recompiles small enough that we can compile them all ahead of time on most renderers.  In particular:
- Metal: Completely eliminates pipeline compiles on all known GPUs
- DX12: Completely eliminates pipeline compiles on the two GPUs I tested (Radeon Pro 5600M with weird AMD Mac bootcamp drivers and a GT 750M with its ancient Kepler Mobile driver)
- Vulkan:
  - Completely eliminates pipeline compiles on MoltenVK on Intel, AMD, and Nvidia GPUs
  - Mostly eliminates pipeline compiles on MoltenVK on Apple GPUs
  - Mostly eliminates pipeline compiles on the two GPUs listed above on their Windows Vulkan drivers

The implementation puts the list of offsets and strides into the vertex uniform, which the shader uses to figure out where it should load from in an SSBO.

I'm not super confident in my Vulkan code and even less in my DX12 code, so if someone could look those over especially closely that would be nice.

<details>
<summary>Before and After recordings</summary>
(Running with all pipeline and driver caches cleared, exclusive ubershaders, compile shaders before starting)

Before:

https://user-images.githubusercontent.com/3315070/175461508-3b744072-e4fd-491c-9aaa-00c83f6295b0.mp4

After (that one stutter is an EFB copy pipeline):

https://user-images.githubusercontent.com/3315070/175463003-7a272875-8bd2-47e1-8afd-e9057c6e028e.mp4

</details>

~~Note: Just waiting on #10747~~All ready

### CODE REVIEW

**Comment by K0bin** [comment]
Wouldn't it make more sense to use VK_EXT_vertex_input_dynamic_state on Vulkan when supported (Nvidia GPUs)?

**Comment by JMC47** [comment]
I'm guessing that's an encoding issue based on the size of the pixels, so I just ignored it.

**Comment by Rumi-Larry** [comment]
I think it would make more sense to make a commit without the Vulkan code

**Comment by tellowkrinkle** [comment]
> Wouldn't it make more sense to use VK_EXT_vertex_input_dynamic_state on Vulkan when supported (Nvidia GPUs)?

Probably, I admittedly wasn't really developing this for Vulkan.  I'm happy to adjust the Vulkan implementation to check `g_ActiveConfig.backend_info.bSupportsDynamicVertexLoader` so that the whole thing can be flipped off by writing `false` into the variable when someone does implement `VK_EXT_vertex_input_dynamic_state` though.  Maybe use a different flag to enable the additional pipelines in the `VideoCommon: Compile a few extra pipelines` commit as well?  Or just always compile them, assuming it'll be really fast on OGL / DX11 since it doesn't actually involve any compilation there?

> I'm guessing that's an encoding issue based on the size of the pixels, so I just ignored it.

I had to compress those pretty hard to get under GitHub's 10mb limit.  Used the same (horribly low) quality settings on both but clearly that doesn't mean x264 will produce similar results.  Can confirm that doesn't happen when you actually play.

> I think it would make more sense to make a PR without the Metal code.

Just to confirm, you think that even though it would still include the spirv-cross PR code?

**Comment by OatmealDome** [comment]
@MayImilae Would be interesting to see if this fixes the problems with frame times you observed on Intel Macs.

**Comment by Pokechu22** [comment]
I haven't looked over this in detail, but it feels like a somewhat strange approach. Dolphin already has a system to reformat vertices from the GameCube's format to one that works well for modern GPUs (`VertexLoaderX64`/`VertexLoaderARM64`), though as you've found it generates different formats based on the original one. IMO it would be better to modify that so that a single vertex with all components is generated. That would mean using 156-byte vertices in all cases (even if the input vertex is just a color and a single texture coordinate), though.

**Comment by tellowkrinkle** [comment]
The majority of vertices I've seen are in the 20-28 byte range, which would mean sending everything would be a 6-8x increase in vertex bytes, just to slightly reduce the complexity of ubershaders.  Messing with stats reporting shows that it would be pretty common to see games hitting 20MB/frame or a bit over 1GB/s of vertices this way.

Side note, if you were hoping to keep vertex loaders out of shaders in general, I was planning on implementing vertex shader line to triangle expansion for platforms without geometry shaders, which will also require a shader vertex loader.

**Comment by JMC47** [comment]
Ah, Line-width without geometry shaders might actually be faster on Android.

File: Source/Core/VideoCommon/ShaderGenCommon.h [hunk 1]
```diff
@@ -302,6 +303,15 @@ static const char s_shader_uniforms[] = "\tuint    components;\n"
...
                                         "\tfloat4 " I_CACHED_BINORMAL ";\n"
+                                        "\tuint vertex_stride;\n"
+                                        "\tuint vertex_offset_rawnormal;\n"
+                                        "\tuint vertex_offset_rawtangent;\n"
+                                        "\tuint vertex_offset_rawbinormal;\n"
+                                        "\tuint vertex_offset_rawpos;\n"
+                                        "\tuint vertex_offset_posmtx;\n"
+                                        "\tuint vertex_offset_rawcolor0;\n"
+                                        "\tuint vertex_offset_rawcolor1;\n"
+                                        "\tuint4 vertex_offset_rawtex[2];\n"  // std140 is pain
```
Comment (Pokechu22) [inline Source/Core/VideoCommon/ShaderGenCommon.h:314]: It seems a bit odd that these are included in the uniform for specialized shaders (and ubershaders that don't use the vertex loader), though I don't know if 64 additional bytes is really a problem.

**Review by Pokechu22 — COMMENTED** [review]
Now that I've looked over it further, this approach doesn't seem that bad (although I think it is a bit jank that we have several things labeled as vertex loaders that all do slightly different things). Since all components are 4-byte aligned, the actual code seems fairly simple, and `uintBitsToFloat` is also convenient.

It might be a bit simpler if there were, say, individual arrays for each component, some of which don't get populated, but I'm not sure if that would work or have any real benefit (and it would require changing the assembly vertex loaders).

> Side note, if you were hoping to keep vertex loaders out of shaders in general, I was planning on implementing vertex shader line to triangle expansion for platforms without geometry shaders, which will also require a shader vertex loader.

Note that there's an existing implementation of this that's rather jank at #10295 (which doesn't use a shader-based vertex loader, but instead duplicates them in `IndexGenerator.cpp`).

File: Source/Core/VideoCommon/UberShaderVertex.cpp [hunk 2]
```diff
@@ -50,15 +54,90 @@ ShaderCode GenVertexShader(APIType api_type, const ShaderHostConfig& host_config
...
+                "uint GetOffset(uint attr_offset) {{\n"
+                "  return (gl_VertexID + base_vertex) * vertex_stride + attr_offset;\n"
+                "}}\n");
+    }
+    else
+    {
+      out.Write("uint GetOffset(uint attr_offset) {{\n"
+                "  return gl_VertexID * vertex_stride + attr_offset;\n"
+                "}}\n");
+    }
```
[the remark this answers is not in the document]
Comment (tellowkrinkle) [reply Source/Core/VideoCommon/UberShaderVertex.cpp:79]: `gl_VertexID` gets translated by spirv-cross to `SV_VertexID`.  `SV_VertexID` starts at zero regardless of what the base vertex was, while OpenGL's `gl_VertexID` (as well as Vulkan's and Metal's equivalents) start at the base vertex.

Do you have any recommendations for things to put into the comment to make it more clear?

File: Source/Core/VideoCommon/ShaderGenCommon.h (hunk 1 above)
Comment (tellowkrinkle) [reply Source/Core/VideoCommon/ShaderGenCommon.h:314]: VS uniform size is currently 4032 bytes, and this makes it 4096 bytes, a ~1.6% increase.

If it becomes a big issue, our full vertex layout can be calculated from the component list plus ~17 bits of component size data (since the order is always the same), just a bit more messily.  We could put something like this at the top of `main` instead:
```glsl
// vs uniform: u32 component_sizes; // bits 0-16: rawtex, bit 17: position has 3 floats
uint vertex_stride = 0;
uint vertex_offset_posmtx;
uint vertex_offset_rawpos;
uint vertex_offset_rawnormal;
uint vertex_offset_rawtangent;
uint vertex_offset_rawbinormal;
uint vertex_offset_rawcolor0;
uint vertex_offset_rawcolor1;
uint vertex_offset_rawtex[8];
if (components & VB_HAS_POSMTX) {
  vertex_offset_posmtx = vertex_stride;
  vertex_stride++;
}
vertex_offset_position = vertex_stride;
vertex_stride += component_sizes & (1 << 17) ? 3 : 2;
if (components & VB_HAS_NORMAL) {
  vertex_offset_rawnormal = vertex_stride;
  vertex_stride += 3;
}
// ... binormal, tangent, color0, color1
for (int i = 0; i < 8; i++) {
  if (components & (VB_HAS_UV0 << i)) {
    vertex_offset_rawtex[i] = vertex_stride;
    vertex_stride += (component_sizes >> (2 * i)) & 3;
  }
}
```
Assuming the compiler recognizes those as uniform, AMD and Intel's ISAs probably wouldn't mind too much, it's just a bunch of sgprs on AMD (we usually have plenty to spare) and would take 2/128 gprs on Intel.  No clue how Nvidia and Apple would fare though, since I'm not sure those two have special storage for non-constant uniform values.

A second question would be how often do we upload new uniforms for something that doesn't affect non-uber shaders.  I would guess this happens more often in PS uniforms than VS uniforms though.  Might still be useful to have separate dirty flags for the two in the future.

**Comment by tellowkrinkle** [comment]
> It might be a bit simpler if there were, say, individual arrays for each component, some of which don't get populated, but I'm not sure if that would work or have any real benefit

Wouldn't that require binding like 15 different vertex buffers?  Or were you thinking of some other way of accessing them all?

File: Source/Core/VideoCommon/UberShaderVertex.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (Pokechu22) [reply Source/Core/VideoCommon/UberShaderVertex.cpp:79]: I wasn't aware of this base index/base vertex behavior beforehand (and I still don't entirely know where the numbers come from, as the `IDX` column in renderdoc isn't continuous between drawcalls; if it's just a buffer that new vertices are put at the end of, I'm not entirely sure how that handles differences in vertex sizes between the drawcalls, since I don't know how they would be aligned (e.g. if you have 9 vertices that are 3 bytes each (so 27 bytes total) and then a 4-byte vertex, does that get located at a 27-byte offset, or 32 bytes?)).

As for the comment itself, something like this would work (although it's a bit wordy):

> Write a function to get an offset into vertex_buffer corresponding to this vertex plus the attribute offset. This must be done differently for D3D compared to OpenGL/Vulkan/Metal, as on OpenGL etc., gl_VertexID starts counting at the base vertex specified in glDrawElementsBaseVertex, while on D3D, SV_VertexID (which spirv-cross translates gl_VertexID into) starts counting at 0 regardless of the BaseVertexLocation value passed to DrawIndexed.

(You could also link to https://registry.khronos.org/OpenGL-Refpages/gl4/html/gl_VertexID.xhtml and https://docs.microsoft.com/en-us/windows/win32/direct3d11/d3d10-graphics-programming-guide-input-assembler-stage-using#vertexid but that's probably excessive)

One other thought: it might be better to have `GetOffset()` return just the offset of the current vertex (e.g. call it `GetVertexOffset()`, and then add the attribute offset outside of it. It would be used something like this:

```glsl
float3 load_input_float3_float3(uint attr_offset) {
  uint vtx_offset = GetVertexOffset();
  return float3(uintBitsToFloat(vertex_buffer[vtx_offset + attr_offset + 0]),
                uintBitsToFloat(vertex_buffer[vtx_offset + attr_offset + 1]),
                uintBitsToFloat(vertex_buffer[vtx_offset + attr_offset + 2]));
}
```

**Comment by Pokechu22** [comment]
>> It might be a bit simpler if there were, say, individual arrays for each component, some of which don't get populated, but I'm not sure if that would work or have any real benefit
>
> Wouldn't that require binding like 15 different vertex buffers? Or were you thinking of some other way of accessing them all?

I don't have an exact idea in mind, but it probably would require binding 15 different buffers, yeah. The main benefit would be reducing the amount of calculation that needs to be done to find the index into those buffers since you would always know the correct buffer and stride for given component. But I'm not sure if that would actually be any faster (or if it's possible to avoid syncing data for buffers that aren't relevant), and it would require lots of other work.

File: Source/Core/VideoCommon/UberShaderVertex.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (tellowkrinkle) [reply Source/Core/VideoCommon/UberShaderVertex.cpp]: > e.g. if you have 9 vertices that are 3 bytes each (so 27 bytes total) and then a 4-byte vertex, does that get located at a 27-byte offset, or 32 bytes?)

Vulkan will align each allocation to its vertex stride: https://github.com/dolphin-emu/dolphin/blob/3f794bf87e4d90193b476cf93247b0e26aedd6a9/Source/Core/VideoBackends/Vulkan/VKStreamBuffer.cpp#L152
Metal does the same.  I haven't checked the others.

> it might be better to have GetOffset() return just the offset of the current vertex (e.g. call it GetVertexOffset(), and then add the attribute offset outside of it. It would be used something like this:

If we go forward with #10890, I'll be changing the signature to
```glsl
float3 load_input_float3_float3(uint vtx_offset, uint attr_offset) {
  uint offset = vtx_offset + attr_offset;
  return float3(uintBitsToFloat(vertex_buffer[offset + 0]),
                uintBitsToFloat(vertex_buffer[offset + 1]),
                uintBitsToFloat(vertex_buffer[offset + 2]));
}
```
to allow fetching nearby vertices' data.  Would you like me to just do that here, or stick with the separate `GetVertexOffset()` function and change it in #10890?

[the remark this answers is not in the document]
Comment (Pokechu22) [reply Source/Core/VideoCommon/UberShaderVertex.cpp]: That signature looks fine to me; please do include that in this PR.

> Vulkan will align each allocation to its vertex stride:

Alright, thanks. That makes sense; I just didn't find that code.

[the remark this answers is not in the document]
Comment (tellowkrinkle) [reply Source/Core/VideoCommon/UberShaderVertex.cpp]: Okay.  I updated it and pushed it to the end for you to look over.  Once you have, say so and I'll squash it in (or make changes if you request that)

[the remark this answers is not in the document]
Comment (Pokechu22) [reply Source/Core/VideoCommon/UberShaderVertex.cpp]: Looks good to me.

File: Source/Core/VideoBackends/Metal/MTLObjectCache.mm [hunk 3]
```diff
@@ -330,10 +331,13 @@ static void CopyAll(std::array<VertexAttribute, N>& output, const AttributeForma
       // Throw extras in bits we don't otherwise use
       if (cfg.rasterization_state.cullmode == CullMode::All)
         blend.hex |= 1 << 29;
-      if (cfg.rasterization_state.primitive == PrimitiveType::Points)
-        blend.hex |= 1 << 30;
-      else if (cfg.rasterization_state.primitive == PrimitiveType::Lines)
-        blend.hex |= 1 << 31;
```
Comment (Pokechu22) [inline Source/Core/VideoBackends/Metal/MTLObjectCache.mm:336]: This is my first time seeing this code specifically and I don't particularly care for that aspect of it.  Doing something like this would be better (and I think even is legal):

```C++
union
{
  BlendingState blend;
  // Throw extras in bits we don't otherwise use
  BitField<29, 1, u32, bool> cull_all;
  BitField<30, 1, u32, bool> is_points;
  BitField<31, 1, u32, bool> is_lines;
}
```

That could be done here or in a different PR (again, assuming it's legal - I don't have a way of testing objective C stuff).

File: Source/Core/VideoBackends/Vulkan/ObjectCache.cpp [hunk 4]
```diff
@@ -123,8 +123,9 @@ bool ObjectCache::CreateDescriptorSetLayouts()
        VK_SHADER_STAGE_FRAGMENT_BIT},
   }};
 
-  static const std::array<VkDescriptorSetLayoutBinding, 1> standard_ssbo_bindings{{
+  static const std::array<VkDescriptorSetLayoutBinding, 2> standard_ssbo_bindings{{
       {0, VK_DESCRIPTOR_TYPE_STORAGE_BUFFER, 1, VK_SHADER_STAGE_FRAGMENT_BIT},
+      {1, VK_DESCRIPTOR_TYPE_STORAGE_BUFFER, 1, VK_SHADER_STAGE_VERTEX_BIT},
   }};
```
Comment (Pokechu22) [inline Source/Core/VideoBackends/Vulkan/ObjectCache.cpp:129]: Doesn't this mean this buffer is always accessible (or something like that), even on non-uber pipelines? `create_infos` always uses `static_cast<u32>(standard_ssbo_bindings.size()), standard_ssbo_bindings.data()`. Or is that handled by the `setLayoutCount--` in `CreatePipelineLayouts`?

File: Source/Core/VideoCommon/ShaderCache.cpp [hunk 5]
```diff
@@ -1275,7 +1278,42 @@ void ShaderCache::QueueUberShaderPipelines()
...
+          //    alphaupdate and colorupdate ones.  Also keyed on primitive type, but Metal supports
+          //    setting it to "unknown" and we do for ubershaders (but MoltenVK won't).
+          // Windows Vulkan:
+          //  - AMD, Nvidia: Definitely keyed on dual source blend, but the others seem more random
+          //    Changing a setting on one shader will require a recompile, but changing the same
+          //    setting on another won't.  Compiling a copy with alphaupdate off, colorupdate off,
+          //    and one with DSB on seems to get pretty good coverage though.
+          // Windows D3D12:
+          //  - AMD: Keyed on dual source blend and vertex layout
+          //  - Nvidia Kepler: Acts just like D3D11
```
Comment (Pokechu22) [inline Source/Core/VideoCommon/ShaderCache.cpp:1301]: Where is this information from? Is it just experimental results, or is it documented somewhere?

(Also, "acts just like D3D11" isn't useful since D3D11 wasn't described anywhere in this comment; what's that referring to?)

**Comment by Pokechu22** [comment]
Other than that things look mostly reasonable, though I don't really understand the D3D or Vulkan setup code. I don't see anything that looks _obviously_ wrong, at least.

File: Source/Core/VideoBackends/Vulkan/ObjectCache.cpp (hunk 4 above)
Comment (tellowkrinkle) [reply Source/Core/VideoBackends/Vulkan/ObjectCache.cpp:129]: Yes, it's always accessible.  We use the same layout object for ubershaders and non-ubershaders, similar to how the GS buffer is included in the layout as long as GS is supported, regardless of whether the current pipeline uses GS or not.

I can remove it in the case where dynamic vertex loader is unsupported though (and probably should, since we don't bind anything there in that case).

File: Source/Core/VideoBackends/Metal/MTLObjectCache.mm (hunk 3 above)
Comment (tellowkrinkle) [reply Source/Core/VideoBackends/Metal/MTLObjectCache.mm]: Sounds good

I removed the cullmode all one while I was at it, since it looks like that gets handled in CPU processing (DX11 backend also doesn't specially handle it)

File: Source/Core/VideoCommon/ShaderCache.cpp (hunk 5 above)
Comment (tellowkrinkle) [reply Source/Core/VideoCommon/ShaderCache.cpp]: Experimental results.  Edited the comment to make it clearer and removed the comparison to DX11.

File: Source/Core/VideoBackends/Metal/MTLObjectCache.mm [hunk 6]
```diff
@@ -220,7 +220,7 @@ static MTLCullMode Convert(CullMode cull)
   switch (cull)
   {
   case CullMode::None:
-  case CullMode::All:  // Handled by disabling rasterization
+  case CullMode::All:  // Handled in CPU
     return MTLCullModeNone;
```
Comment (Pokechu22) [inline Source/Core/VideoBackends/Metal/MTLObjectCache.mm:224]: Specifically, it's done here:

https://github.com/dolphin-emu/dolphin/blob/d04e1ca96a705f606d5ba75f3bb915a0ea84571a/Source/Core/VideoCommon/VertexLoaderManager.cpp#L358-L362

I'd write this as "Handled by VertexLoaderManager::RunVertices" instead of just "in CPU".

I'm not sure why that code doesn't skip lines and points; this might be a hardware quirk (since I don't think lines/points can be drawn in such a way that backface culling applies, and they might not be culled at all; they're not culled in the software renderer as well).

In any case, the only pipeline that _does_ implement `CullMode::All` on the GPU is Vulkan (although apparently `VK_CULL_MODE_FRONT_AND_BACK` it [only applies to triangles](https://books.google.com/books?id=XVBxDQAAQBAJ&pg=PT338&lpg=PT338&q=VK_CULL_MODE_FRONT_AND_BACK)).

<details><summary>Relevant backend code</summary>

https://github.com/dolphin-emu/dolphin/blob/d04e1ca96a705f606d5ba75f3bb915a0ea84571a/Source/Core/VideoBackends/OGL/OGLRender.cpp#L1125-L1135
https://github.com/dolphin-emu/dolphin/blob/d04e1ca96a705f606d5ba75f3bb915a0ea84571a/Source/Core/VideoBackends/D3D/D3DState.cpp#L441-L442
https://github.com/dolphin-emu/dolphin/blob/d04e1ca96a705f606d5ba75f3bb915a0ea84571a/Source/Core/VideoBackends/D3D12/DX12Pipeline.cpp#L64-L66
https://github.com/dolphin-emu/dolphin/blob/d04e1ca96a705f606d5ba75f3bb915a0ea84571a/Source/Core/VideoBackends/Vulkan/VKPipeline.cpp#L43-L45

</details>

Comment (tellowkrinkle) [reply Source/Core/VideoBackends/Metal/MTLObjectCache.mm]: > I'd write this as "Handled by VertexLoaderManager::RunVertices" instead of just "in CPU".

Done

> I'm not sure why that code doesn't skip lines and points; this might be a hardware quirk (since I don't think lines/points can be drawn in such a way that backface culling applies, and they might not be culled at all; they're not culled in the software renderer as well).

Yeah lines/points will have their cull mode set to none before getting sent to the render backend (which is convenient for anyone who might be trying to expand them to triangles on the GPU)
https://github.com/dolphin-emu/dolphin/blob/d04e1ca96a705f606d5ba75f3bb915a0ea84571a/Source/Core/VideoCommon/RenderState.cpp#L16-L18

**Review by Pokechu22 — APPROVED** [review]
Looks good to me. I don't 100% understand everything, but I think it's good enough to merge for users to find any remaining subtle bugs.
